common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt

```diff
@@ -275,70 +262,12 @@ internal class WebSocketCoreImpl(
-                        }
-                    }
-                    awaitClose {
-                        wsScope.launch {
-                            var subscription: Long? = null
-                            eventSubscriptionMutex.withLock {
-                                synchronized(activeMessages) {
-                                    activeMessages.entries.firstOrNull { it.value.message == subscribeMessage }
-                                        ?.let {
-                                            subscription = it.key
-                                            channel.close()
-                                            activeMessages.remove(subscription)
-                                        }
-                                }
-                            }
-                            subscription?.let {
-                                Timber.d("Unsubscribing from $type with data $data")
-                                sendMessage(
-                                    mapOf(
-                                        "type" to "unsubscribe_events",
-                                        "subscription" to it,
-                                    ),
-                                )
-                            }
-                            if (activeMessages.isEmpty()) {
-                                Timber.i("No more subscriptions, closing connection.")
-                                connection?.close(1001, "Done listening to subscriptions.")
-                            } else {
-                                Timber.i("Still ${activeMessages.size} messages in the queue, not closing connection.")
-                            }
-                        }
-                    }
-                }.shareIn(backgroundScope, SharingStarted.WhileSubscribed(timeout.inWholeMilliseconds, 0))
-
-                val webSocketRequest = WebSocketRequest(
-                    message = subscribeMessage,
-                    eventFlow = flow,
-                    onEvent = channel,
-                )
-                val response = sendMessage(webSocketRequest)
-                if (response == null || response.success != true) {
-                    Timber.e("Unable to subscribe to $type with data $data")
-                    synchronized(activeMessages) {
-                        activeMessages.entries
-                            .firstOrNull { it.value.message == subscribeMessage }
-                            ?.let { activeMessages.remove(it.key) }
-                    }
-                    return null
-                }
-            }
+        return eventSubscriptionMutex.withLock<Flow<T>?> {
+            ( // Check for existing subscription before creating a new one
+                activeMessages.entries.firstOrNull { it.value.message == subscribeMessage }?.value?.eventFlow
+                    ?: createSubscriptionFlow<T>(subscribeMessage, timeout)
+                ) as? Flow<T>
```

Comment:
Nothing changed here except the removal of the synchronized block. I've made it this way to avoid iterating multiple times on the activeMessages.